[Orders] Ability to filter by product

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #12059 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

Allows merchants to view orders by filtering them using a product. 

- Adds a new row to select the product for filtering orders
- The selected product info is saved to settings and restored upon app relaunch. 


## Testing instructions

- Create a store
- Create a few products and orders
- Login into the app
- Navigate to the Orders tab
- Tap "Filter" -> "Product" 
- From the list of products select one
- Tap "Show Orders"
- The order list screen should show only the orders related to the selected order now.
- Tap "Filter" again and validate that the previously selected product name is visible in the filter.
- Tap "Clear all" to clear the filters. 
- Applying the filter should start showing all the orders in the order list screen.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/524475/1f87f218-58fe-4fae-920b-7fb5d47def1f



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Update order syncing with an optional product ID

selanthiraiyan committed:
Support filtering first page of the order list.

selanthiraiyan committed:
Add configuration for product selector.

selanthiraiyan committed:
Handle order filter case in product list selector.

selanthiraiyan committed:
Add new filter for filtering orders by product.

selanthiraiyan committed:
Store and retrieve product filter settings.

selanthiraiyan committed:
Show product filter row to select product.

selanthiraiyan committed:
Use selected product to filter orders.

selanthiraiyan committed:
Inject product filter in unit tests.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Wrap product selector in a navigation controller.

selanthiraiyan committed:
Fix linting errors.

selanthiraiyan committed:
Update product filter selection screen's title.

selanthiraiyan committed:
Remove unused variable.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Filters/FilterListViewController.swift:
```diff
@@ -248,6 +250,34 @@ private extension FilterListViewController {
+                        selectedItemIDs: selectedProductID,
+                        onProductSelectionStateChanged: { product in
+                            selected.selectedValue = FilterOrdersByProduct(id: product.productID, name: product.name)
+                            self.updateUI(numberOfActiveFilters: self.viewModel.filterTypeViewModels.numberOfActiveFilters)
+                            self.listSelector.reloadData()
+                            self.listSelector.navigationController?.dismiss(animated: true)
```
We should capture `self` weakly here.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12089/commits/97a662a6c51ad239ea00d705aad4f3dd605da1f2

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Filters/ProductSelector+Order.swift:
```diff
@@ -0,0 +1,33 @@
+        title: Localization.productSelectorTitle,
+        cancelButtonTitle: Localization.productSelectorCancel,
+        productRowAccessibilityHint: Localization.productRowAccessibilityHint,
+        variableProductRowAccessibilityHint: Localization.variableProductRowAccessibilityHint
```
Nit: this should not be needed since we're not opening variation lists.

selanthiraiyan replied:
`variableProductRowAccessibilityHint` is a non-optional value in `ProductSelectorView.Configuration`, and we need to specify this in the configuration. I think changing this behaviour will be outside the scope of this PR.

itsmeichigo commented on Yosemite/Yosemite/Model/Orders/StoredOrderSettings.swift:
```diff
@@ -27,6 +30,9 @@ public struct StoredOrderSettings: Codable, Equatable {
             if let dateRangeFilter = dateRangeFilter, dateRangeFilter.filter != .any {
                 total += 1
             }
+            if let _ = productFilter {
```
Nit:
```suggestion
            if productFilter != nil {
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/12089/commits/b677b9837fa63f29e59f1b19cbd275e6d7860fbb

selanthiraiyan committed:
Adds feature flag for enabling filtering orders by product.

selanthiraiyan committed:
Show product filter row only if the feature flag is turned on.

selanthiraiyan committed:
Merge branch 'trunk' into feat/12059-product-filter

selanthiraiyan committed:
Capture self weakly.

selanthiraiyan committed:
Use nil to check condition.

selanthiraiyan committed:
Add file to hold analytics events for orders filter.

selanthiraiyan committed:
extension to calculate analytics value.

selanthiraiyan committed:
Use nil as analytics value if no date range selected.

selanthiraiyan committed:
Helper to calculate analytics value for product filter.

selanthiraiyan committed:
Use helper method to track order filter event.

selanthiraiyan commented:
Thanks for the review, Huong!

- I addressed your comments and also pushed a few more changes to update the existing `orders_list_loaded` tracks event and started tracking the selected product ID. 
- I also made some refactoring to avoid tracking empty strings when there is no order status or date range filter selected. This helps to avoid having empty strings in `orders_list_loaded` tracks events. 

The PR is ready for another look. 🙇

itsmeichigo reviewed (approved):
LGTM :shipit:
